rfcs/20230704-dynamism-101.md

```diff
@@ -0,0 +1,825 @@
+was a big help. Referring to this spreadsheet will likely help in reviewing this
+RFC as well.
+
+Before beginning, let's align on terminology. In discussions around MLIR, it is
+fairly conventional to use the word "dimension" ambiguously - to refer to either
+an actual dimension or the size of a dimension. Most of the time, the exact
+meaning of the word "dimension" is clear from the context, but sometimes it's
+ambiguous. For example, it is not obvious whether
+`mlir::ShapedTypeComponents::getDims` returns dimension numbers or dimension
+sizes (for what it's worth, it is the latter - it returns dimension sizes).
+
+To avoid ambiguity, this document will be using the following terminology:
+
+* **Dimension numbers** (or **axes**) to refer to actual dimensions,
+  e.g. "`tensor<16x?xf32>` has two axes".
+* **Dimension sizes** (or **sizes**) to refer to sizes of dimensions,
+  e.g. "`tensor<16x?xf32>` has the following dimension sizes: 16 and unknown".
+* Unqualified "dimension" will not be used at all.
+
+By the virtue of being bootstrapped from MHLO, the StableHLO dialect already
+has support for **dynamism within types**:
+
+* **Unbounded dynamism**: In a tensor type, some or all dimension sizes may be
+  unknown (aka "dynamic"). In MLIR syntax, these dimension sizes are expressed
+  via question marks.
+* **Bounded dynamism**: The same but some of dynamic dimensions sizes may have
+  known upper bounds. In MLIR syntax, these dimension sizes are expressed via
+  question marks, and bounds are expressed via `#stablehlo.bounds`.
+* **Unranked dynamism**: In a tensor type, rank may be unknown. In MLIR
+  syntax, this fact is expressed via asterisks.
+
+```mlir
+// Static shapes:
+// All dimension sizes are known.
+%0 = stablehlo.add %arg0, %arg1 : tensor<16x16xf32>
+
+// Unbounded dynamism:
+// First dimension size is unknown (?).
+// Second dimension size is known (16).
+%1 = stablehlo.add %arg0, %arg1 : tensor<?x16xf32>
+
+// Bounded dynamism:
+// First dimension size is unknown (?), but its bound is known (16).
+// Second dimension size is known (16), so its bound is N/A (?).
+%2 = stablehlo.add %arg0, %arg1 : tensor<?x16xf32, #stablehlo.bounds<16, ?>>
+
+// Unranked dynamism:
+// The rank is unknown (*).
+%3 = stablehlo.add %arg0, %arg1 : tensor<*xf32>
+```
+
+Similarly, the StableHLO dialect already has support for **dynamism within
+operations**, with some ops such as PadOp having dynamic counterparts such as
+DynamicPadOp. For example:
+
+```mlir
+// "vanilla" PadOp:
+// low, high and interior paddings are implemented via MLIR attributes,
+// i.e. they are static.
+%0 = stablehlo.pad %arg0, %arg1, low = [1, 1], high = [1, 1], interior = [0, 0]
+  : (tensor<16x16xf32>, tensor<f32>) -> tensor<20x20xf32>
+
+// DynamicPadOp:
+// low, high and interior paddings are implemented via MLIR operands,
+// i.e. they are dynamic.
+%1 = stablehlo.dynamic_pad %arg0, %arg1, %arg2, %arg3, %arg4 :
```

Comment:
nit: Would be clearer if you gave these args names like "low", "high", "interior" etc

Reply:
Agreed - we have a series of tickets to improve prettyprinting for StableHLO, but they have a relatively lower priority than the work that's on the critical path for v1.0. https://github.com/openxla/stablehlo/blob/main/docs/roadmap.md goes into details about the overall roadmap.

Reply:
Ack that's OK, I just meant in this proposal in this actual text on line 101. Instead of arg0, arg1, etc, change line 101 to be:

%1 = stablehlo.dynamic_pad %tensor, %low, %high, etc